File: fms_core/src/main/java/net/com/fms_core/script/geo_service.py
import sqlite3
import requests
import time

DB_PATH = "geo_cache.db"


class GeoService:

    def __init__(self):
        self._create_table()

    def _create_table(self):
        conn = sqlite3.connect(DB_PATH)
        cur = conn.cursor()
        cur.execute("""
            CREATE TABLE IF NOT EXISTS geo_cache (
                location TEXT PRIMARY KEY,
                latitude REAL,
                longitude REAL,
                updated_at INTEGER
            )
        """)
        conn.commit()
        conn.close()

    def get_coordinates(self, location):
        cleaned = self._clean(location)
        cached = self._get_cached(cleaned)

        if cached:
            return cached

        coords = self._query_api(cleaned)

        if coords:
            self._save(cleaned, coords)
            return coords

        return None

    def _clean(self, text):
        if not text:
            return ""
        text = text.strip()
        text = " ".join(text.split())
        return text

    def _get_cached(self, location):
        conn = sqlite3.connect(DB_PATH)
        cur = conn.cursor()
        cur.execute("SELECT latitude, longitude FROM geo_cache WHERE location = ?", (location,))
        row = cur.fetchone()
        conn.close()

        if row:
            return row[0], row[1]
        return None

    def _save(self, location, coords):
        conn = sqlite3.connect(DB_PATH)
        cur = conn.cursor()

        cur.execute(
            "REPLACE INTO geo_cache (location, latitude, longitude, updated_at) VALUES (?, ?, ?, ?)",
            (location, coords[0], coords[1], int(time.time()))
        )

        conn.commit()
        conn.close()

    def _query_api(self, location):
        try:
            url = "https://nominatim.openstreetmap.org/search"
            params = {
                "q": location + ", Sri Lanka",
                "format": "json"
            }
            headers = {"User-Agent": "GeoLookupScript"}

            res = requests.get(url, params=params, headers=headers, timeout=10)
            data = res.json()

            if not data:
                return None

            lat = float(data[0]["lat"])
            lon = float(data[0]["lon"])
            return (lat, lon)
        except Exception:
            return None

File: fms_core/src/main/java/net/com/fms_core/service/impl/GeoService.java
package net.com.fms_core.service.impl;

import com.fasterxml.jackson.databind.ObjectMapper;
import lombok.RequiredArgsConstructor;
import net.com.fms_core.dto.ImpossibleDistanceResult;
import net.com.fms_core.dto.message.IsoMessageDTO;
import net.com.fms_core.entity.geo.GeoCache;
import net.com.fms_core.repository.geo.GeoCacheRepository;
import org.springframework.stereotype.Service;
import org.springframework.web.client.RestTemplate;

import java.util.Map;

/**
 * ALL RIGHT RESERVED By kavinda_d
 *
 * @AUTHOR : kavinda_d
 * @PROJECT : fms backend
 */

@Service
@RequiredArgsConstructor
public class GeoService {

    private final GeoCacheRepository repo;

    public double[] getCoordinates(String location) {

        String cleaned = clean(location);
        GeoCache cached = repo.findById(cleaned).orElse(null);

        if (cached != null) {
            return new double[] { cached.getLatitude(), cached.getLongitude() };
        }

        double[] coords = queryApi(cleaned);

        if (coords != null) {
            GeoCache g = new GeoCache();
            g.setLocation(cleaned);
            g.setLatitude(coords[0]);
            g.setLongitude(coords[1]);
//            g.setUpdatedAt(System.currentTimeMillis());
            repo.save(g);

            return coords;
        }

        return null;
    }

    private String clean(String text) {
        if (text == null) return "";
        return text.trim().replaceAll(" +", " ").toLowerCase();
    }

    // Common country code aliases to normalize before geocoding
    private static final Map<String, String> COUNTRY_ALIASES = Map.of(
        "sl", "Sri Lanka",
        "lk", "Sri Lanka",
        "us", "United States",
        "uk", "United Kingdom",
        "gb", "United Kingdom",
        "in", "India",
        "au", "Australia"
    );

    private double[] queryApi(String location) {
        try {
            // Parse location: "MERCHANT NAME     City     CountryCode"
            String[] parts = location.trim().split("\\s+");
            String city = "";
            String country = "";

            if (parts.length >= 2) {
                String rawCountry = parts[parts.length - 1].toLowerCase();
                country = COUNTRY_ALIASES.getOrDefault(rawCountry, parts[parts.length - 1]);
                city = parts[parts.length - 2];
            } else {
                city = location;
            }

            String query = city + (country.isEmpty() ? "" : ", " + country);
            String encodedQuery = java.net.URLEncoder.encode(query, "UTF-8");
            String url = "https://nominatim.openstreetmap.org/search?format=json&q=" + encodedQuery;

            RestTemplate rest = new RestTemplate();
            rest.getInterceptors().add((request, body, execution) -> {
                request.getHeaders().add("User-Agent", "FMS-FraudDetection/1.0");
                return execution.execute(request, body);
            });

            var response = rest.getForObject(url, Object[].class);

            if (response == null || response.length == 0) {
                return null;
            }

            Map map = (Map) response[0];
            double lat = Double.parseDouble(map.get("lat").toString());
            double lon = Double.parseDouble(map.get("lon").toString());

            return new double[] { lat, lon };

        } catch (Exception e) {
            return null;
        }
    }
}
